apps/web/pages/_app.js

```diff
@@ -0,0 +1,51 @@
+// @flow
+
+import React from 'react';
+import App, { Container, type AppInitialProps } from 'next/app';
+import {
+  Provider as NavigationProvider,
+  Consumer as NavigationConsumer,
+} from '@kiwicom/margarita-navigation';
+import isEqual from 'react-fast-compare';
+
+import Layout from '../components/Layout';
+
+export default class MyApp extends App {
+  static async getInitialProps({ Component, router, ctx }: AppInitialProps) {
+    let pageProps = {};
+    if (Component.getInitialProps) {
+      pageProps = await Component.getInitialProps(ctx);
+    }
+
+    let params;
```

Comment:
Thanks. Just played with it for a while and for now it seems to cause some weird fetch errors (HTTP error 400) as seen in console and crashes during redirect from `Search` to `Results` (Maximum update depth exceeded). Probably because `hydrateSelf` current hacky approach that you have mentioned. 
But I think that if we clean it a bit this can be the way to do it.